### ISSUE TITLE
4/16 SIM UI Design Review and Development Policy and Process - DRAFT

### ISSUE THREAD

**Issue by jamesmrollins** [lzim/teampsd#921]
**1. Add description:** 

**1.0 Problem**
Projects are getting bottled up on the test server.  Consider multiple TEST project instances in Epicenter.  The intent is to create pathways from DEV to PROD that have different velocities.  For example, if a small bug occurs and needs to be corrected, it can go through DEV to a TEST-Fast instance and straight through to PROD.  However, if there is a more lengthy project that may require user testing and such, these projects can go on a TEST-Slow instance.  

**2.0 Facts Bearing on the Problem**
- Takes about 8 hours of DEV time to set up a new test instance.
- "i"information, portraits, sliders and icons are all dynamic data stored on Epicenter. These may be the source of future version control problems.
- Having a separate instance for a given project will make user testing somewhat easier, as we can point people to a given login and get feedback.  
- PLDR implementation would help with user feedback.
- Keeping the instances up-to-date with each other will cost additional development hours.  
- Estimates are difficult, as the amount of time is proportional the size of the development gap between instances.

_Notes and Action Items_
- If we're on top of maintenance, could take 2 DEV hours to keep instances aligned, if not could take 8 hours.
- Version control needs to be more precisely implemented. Versioning was switched to an hourly system. Drifted away from 1.6 to 1.7 mentality, we should go back to that if we move forward with this.
- SIM UI data is populated by individual teams via team data files. Stored in epicenter. Unless the models themselves change we won't have large data  versioning issues.

**3.0 Risks**
- Epicenter uses many outside files stored in directories in the Dynamic Data folder. There is a small risk they could fall behind in version control.  
- Projects that are not properly "sized up" may plug up the works.  Sometimes, schedule availability of team members will hold up a project for months.  Furthermore, shifting design intent will cause further delays.
- Depending on the dwell time between Test instances, some instances may get behind others.  For example, say Tutor and PLDR are on their own instances.  PLDR is released for PROD.  Now the Tutor instance is behind because it doesn't have the PLDR function.  

_Notes and Action Items_
- Build out a test algorithm, to streamline and formalizing the testing process. 
- Outline which workgroups should be testing a screen and simplified manner of reviewing.
- Without a testing methodology with limits, incentives and feedback we would perhaps run into endless feedback loops that would defeat the goals of agile development.
- We need to put things in front of stakeholders sooner. Ideally before development is made if we want to be Agile.

**4.0 Courses of Action (COA)**
_Note:  In all courses of action, strongly recommend moving "i" information file to the data API and controlling inputs via an admin panel.  Estimate of hours to complete found below in the estimate._

**COA 1**  Leave everything as it is and get better at agile project management. Define sprints to last two weeks and provide _effective_ cross functional support for the sprint.  In other words, when someone is on a sprint, the sprint activities take the highest priority. Processing bugs can move straight to PROD, since the scale of their change is very small.  _Note:  Getting rid of "fuzzy icons" although sounds like a small bug, is actually a rather large undertaking because of the high number of instances of the feature in the code. Therefore, would really qualify for a sprint._

 **COA 2** Create two TEST instances.  One called "Fast" and one call "Slow."  Fast is for bugs and related items that take under 8 development hours to correct.  Slow is for major version control releases.  Slow will only take one project at a time, to maintain the design-teams' focus on completing one thing before starting another.

**COA 3** Create a TEST instance for each "Slow" project that comes to life.  For example, in our current lay down of features, we would have two TEST instances; one for PLDR and one for Tutor.  This would allow for multiple projects moving forward at a time, at different rates.  It would give us the ability to abandon a given project, or to shift resources away from it. However, it could also result in a proliferation of instances that will be costlier to keep up-to-date with each other.

_Notes and Action Items_
- Definition of a small bug should be 8 or less hours of DEV time
- Create 2 TEST instances and improve the adherence to agile principles.  Meaning, we should absolutely manage development with a specific design-development-test procedure. 

**5.0 Discussion**
1.  With COA 3, Because there are no significant or distinct operational branches to the SIM UI (it is pretty much a single purpose app) like finance or administration with unique features, it is not cost-effective to manage an instance for each project in the SIM UI. 
2.  Managing the instances described in COA 3 will become a bottle-neck in and of itself.
3.  COA 1 does not have a "fast" branch, which could still be a bottle  neck.
4.  Need to ensure that an rubric for determining cost-benefit to help inform decisions.

**6.0 Recommendation**

We recommend COA 2 concurrent with the development of a design-sign-off-testing policy and procedure to maintain agile flexibility in development functions.

**Proposed Design and Implementation Policy Outline**

1.  Requirements Definition, Traceability, Validation and Verification
- Requirements Traceability and Verification Matrix (RTVM) development
- Link from requirements to design requirement to testing

2.  Mock-ups, Prototypes and Design Reviews.

3. Test Case Development, Stakeholder Identification and Testing
- Criteria for feedback versus new feature adjudication
- Non functional requirements (NFR) adjudication (i.e., disaster recovery plan)

 **2. Click on the right:**
a. **Projects** - assign to **issue_tracker.**
b. **Assignee** - assign relevant team members.
c. **Labels**- review and select all potential _MTL_ dependencies.
d. **Milestones** - select the dependent milestone.

-----------------------------------------------------------------------------------------
**Workgroups leads**
**Identify Constraints (check all that apply):**
- [ ]   Capacity
- [ ]   Cost
- [ ]   Schedule
- [ ]   Interference
- [ ]   Capability
- [ ]   Other (describe)

_Add estimated person-hours to complete:_ 
- 8 Hrs DEV for each TEST instance.
- 6 Hrs PM to write policy and procedure
- 17 Hrs. to add "i" Information admin functionality.

**Edit the due date (if necessary)**  

### DUE 28 FEB

- 6 FEB:  First draft review with Anthony, Ash and James. 
- 7 FEB: Test instance set up by DEV.
- 10 FEB: Presentation to PI and Workgroup Leads.
- 10 - 17 FEB: Policy review period.
- 17 - 28 FEB: Catch up of Tutor function related development to make consistent with policy (with respect to development of RTVM and testing process.)

**Comment by lzim** [comment]
@lzim will review this issue on Wednesday 2/5

**Comment by lzim** [comment]
@lzim 

**Definitions/Clarity about the alignment of these terms is still needed:**
1. "Instances" how many really needed for "Fast" and "Slow" 
2. "Cost" per "instance" "project"
3. "Maintenance" who does this? 
4. "Versioning" "releases" (e.g., MTL 1.6 , 1.7) versus DevOpps, Subversion, etc.
5. "Test" schedule risk - aren't some things tested by Support Workgroups, and others need other Workgroups like Facilitate Workgroup (this should be clarified in the Jan 2020 update of the SOP)
6. "Bug" definition of <8 hours is proposed.

**Next Steps**
1. Triage > Decision > Issue (Fast) [Issue Tracker] > Feature (Slow) [Milestone timeline to hand-off dependencies]

**Comment by lzim** [comment]
**Summary**
1. We have updated the overall Team PSD SOP to better address monthly milestones and dependencies for moving features through dependent workgroup testing within existing meetings during the month
- Need to consider the overall "RTVM" better pin down the connection between the SOP and this Sim UI Development Policy and Process.
2. We will have a need to move things through "fast"
3. We will always have things that need to move "slow" for the reasons discussed above.

**Proposal is:**
- Create "TEST Fast" now with up to 8 hours of DEV time.
- Move whatever we can to TEST FAST and PROD.
- Work on the "RTVM" connections between the larger Team PSD SOP and the Sim UI Workgroup Design and Production Policy (using the PLDR as a helpful use case).

**Comment by anthonycpichardo** [comment]
# Discussions from 2020_02_07  Pull up

**Will not be presenting first draft on 2/10, will push back so we have time to review the content**

1. ## Define a meeting rhythm
     - Goal: improve timing of project releases and allow for appropriate scheduling
     - Does not need to be rigid as we need to allow for flexibility
     - Meetings should fall in line with milestones set at the first Monday meeting
2. ## Lock Requirements and Create Test Cases
     - Define a point in the project life cycle where we can no longer change requirements if we want to release on schedule.
     - Test Cases should be built around those requirements
3.  ## Quantative and Testing workflows
     - This policy will impact the overall workflow, need to define how other pieces will fall in line
     - Test and Quantitative teams will need a similar process to follow
     - Ash and Anthony will work on this

**Comment by jamesmrollins** [comment]
Hi @anazariz @anthonycpichardo @staceypark @branscombj  - Below is the DRAFT Requirements Design and Development Procedure.  Feedback encouraged - Enjoy!  ~ James

[req_design_dev_procedure.docx](https://github.com/lzim/teampsd/files/4183549/req_design_dev_procedure.docx)

**Comment by anthonycpichardo** [comment]
This looks good @jamesmrollins! I really think this process will be super helpful.

One thing to note will be any additions you might want to include with the integration of zenhub to our github workflow. Nothing major will really be changing so it may not be necessary--will be up to your discretion. We should be disseminating more info once we have the workspace set up

**Comment by staceypark** [comment]
@jamesmrollins was the document updated in relation to Anthony's comment here?

> This looks good @jamesmrollins! I really think this process will be super helpful.
> 
> One thing to note will be any additions you might want to include with the integration of zenhub to our github workflow. Nothing major will really be changing so it may not be necessary--will be up to your discretion. We should be disseminating more info once we have the workspace set up

**Comment by jamesmrollins** [comment]
Actions and edits from 4/2 meeting:

- [ ] 1. Update diagram so that Quant WG has a similar Standard Design Life Cycle (SDLC) process similar to Sim UI.
- [ ] 2. Reinforce the RTVM Sign Off as a requirement in the procedure (design review 1).
- [ ] 3. Regarding the integration of the project and ZenHub trackers.
   -  A single or "master" card will be used for each project and tracked by issue number.  Other major task cards can be used to create more granular work breakdown, such as between design, dev and test steps at the discretion of the project owner.  All cards will require an assignee, cross-functional labels as required, pipeline assignment, milestone assignment, and an level of effort estimate.
   -  Feature - will flow from Triage to Work Breakdown.  Once approved it will rest in Operations To Do in the feature_Tracker.  It will go into in_progress when it is promoted from ice_box, through, Epic backlog, to upcoming epic then Priority Epic.  When in development, it will be moved to In Progress in ZenHub and under_develoment in the feature_tracker.  When product is in functional_testing in feature_tracker, then card will be moved to Review QA in ZenHub. When testing and review complete, cards are moved to Done in both trackers.
   - Bugs typically don't require design documentation.
- [ ]  4. Add SDLC to lexicon.
- [ ]  5. Propose versioning system and criteria.
- [ ]  6. Provide a spreadsheet version of testing script that includes:

   - A method test steps by requirements number (strengthen RTVM tie-in).
   - Use dropdowns to control quality of inputs.
   - Integrate quality information (such as problem codes and root cause codes)

**Comment by anthonycpichardo** [comment]
@lzim @jamesmrollins @anazariz.

Ash and I will be drafting a proposal for the workflow across pipelines in the project trackers and zenhub. 
This will address step 3. in the checklist above and outline what each projects header card should look like.

**Comment by lzim** [comment]
Hi @jamesmrollins @anthonycpichardo @anazariz 

Following up on this 😸  
I don't see the details of my meetings with @anazariz and @anthonycpichardo recorded in @jamesmrollins post above. 📝 

I see in the **Workgroup Leads Meeting** today 4/6 that #921 was described in **sim_ui_workgorup New Business** as a @jamesmrollins continuing to work on it, which he has been since January.
- But, I also see it as top priority for **quant_workgroup new business**I'm assuming this message has been communicated by @anthonycpichardo and @anazariz outside the Lucid Agenda to @jamesmrollins 💬 
- Note: It is preferable for workgroup member **assignees** (not the PI) to outline their understanding of requirements, whenever possible, to confirm understanding and conserve PI time.

My instructions to @anthonycpichardo & @anazariz since last Wednesday 4/1 have been for them to _suspend all other efforts_ until the following is completed ✅ 
A) SOP is finalized.
B) SOP is used to fill out all **quant_workgoup** open issues.
C) Open issues are efficiently reviewed in a **meeting scheduled** with me to be sure all requirements are adequately & accurately _specified_ & _estimated._ 

I re-iterated this instruction on Friday 4/3, when I learned that @jamesmrollins was once again moving this task forward, which does not align with our need to make sure **quant_workgroup** is on the right track with priorities as we make COVID-19 pivots ⚡️ 

**Support Workgroup Leads** on Wednesday 3/25 - I asked @anthonycpichardo to take the lead finalizing this SOP ASAP (cross-ref #1218) with the following goals:
- [ ] 1. **Consistency** with the rest of existing Team PSD SOP
- [ ] 2. **Unique** requirement and testing needs for the **bug_tracker** and **feature_tracker**
- [ ] 3. **Develop** the resources that will be used in terms of a **checklist** and/or **map** for how a **master bug** or **master feature**

Last Friday, 4/3, #hq discussed the importance with @anthonycpichardo and @anazariz to **_stay consistent with existing trackers_ and their workflow**, such as the **manuscript_tracker** and **document_tracker** in the following dimensions (some of which are in @jamesmrollins comment above, some not):
- [ ] 1.  A **master card** for the resource (feature/bug) 
- [ ] 2.  The Master card uses a **standard checklist** that determines advancement through the tracker and **_can be viewed at a glance in the tracker_** (like manuscript_tracker, where each # in the checklist defines tasks for moving from column to column in the tracker/kanban)
- [ ] 3. **Links provided for the Issue #** for any issue task cards (story points) describing effort relevant to that master (analogue to document tracker).

@anthonycpichardo and @anazariz - **As we discussed Friday, here is your top priority for today 4/6:** 📆 
- [ ] 1. Finalize **Master Issue Card** template for the Bug and Feature Tracker by #hqhuddle mid-day for review by #hq.
- [ ] 2. Finalize **quant_workgoup** master card requirements by COB.
- [ ] 3. Re-schedule the post-poned **quant_workgroup** meeting for tomorrow at 9AM Pacific to review these requirements.

Based on our meetings about this and my prior review of the draft SOP, I would assume the following (at a minimum) would be detailed 💫 

:dart:
**Bug & Feature Requirements Trace and Verification Master Card**

**1. Requirement Definition (What & Why)**

**2. Verification Method (How)**
- **Analysis** - code, data, model 
- **Test** - Procedures must be well documented
- **Inspect** - purpose, methods and procedures must be well documented so result can be duplicated using the same inspection procedures
- **Demonstrate** - A comphrehensive demonstration and/or application of the functions and features of the system or its components
- **Infer** from “Why”/purpose of requirement definition.

**3. Reviewer (Who)**
- Within Workgroup Member
- Verify across workgroups
- _Required or Optional?_

**4. Location in Tracker**  
- Determine rules for advancement through kanban defined by checklist.

**All - Other comments I gave to @anazariz and @anthonycpichardo that I do not see reflected in the last post above in this thread** ✍️ 
- Please note that we cannot put Excel Files in GitHub from any VA Machine, whereas a checklist on GitHub can be reviewed via the tracker (see **manuscript_tracker** and **document_tracker**)
- Please _do NOT add new language to the lexicon_ unless absolutely essential. It is far preferable to avoid acronyms and to use existing terms like a "master card" and "issue task card" or "tracker" and not a life cycle.

Thanks!

**Comment by lzim** [comment]
> @lzim @jamesmrollins @anazariz.
> 
> Ash and I will be drafting a proposal for the workflow across pipelines in the project trackers and zenhub.
> This will address step 3. in the checklist above and outline what each projects header card should look like as well as the content in the issue cards themselves

Great thanks @anthonycpichardo! I recorded the details of our last two weeks of meetings, as I did not see them reflected here yet 📑

**Comment by anthonycpichardo** [comment]
@lzim 
>Inspect - purpose, methods and procedures must be well documented so result can be duplicated using the same inspection procedures
Demonstrate - A comprehensive demonstration and/or application of the functions and features of the system or its components
Infer from “Why”/purpose of requirement definition.

Not sure I tracked what you're looking to see for the Inspect and Infer sections. Can you elaborate on what you're looking for here?

For the demo this will likely be optional depending on whether or not its something that can actually be showcased.

We will have the template ready for the huddle.

**Comment by ritahitching** [comment]
**Process for Documentation of a Feature**
1.Decide if the request is a:
(i) Feature
(ii) Bug
(iii) Task
2. Open the commensurate Master Card

**Comment by ritahitching** [comment]
To Do
Change the language to Powerpoint Slide to be consistent with TeamPSD Language

**Comment by lzim** [comment]
@anazariz @anthonycpichardo 

### Example Requirement 1
#### Bug & Feature Requirements Trace and Verification Master Card

1. Requirement 1 Definition (What & Why)
Community Button links out of Sim UI to the GitHub Community of Practice.
2. Verification Method (How)
- [X] Documentation of requirements - Reviewed definition of the requirements at **sim_UI_workgroup** on /4/07/2020.
3. Reviewer (Who & When)
- [X] Verified with Lindsey & Stacey at HQ
4. Location in Tracker
_Determine rules for advancement through kanban defined by checklist._

### Example requirement 2
#### Bug & Feature Requirements Trace and Verification Master Card

1. Requirement 2 Definition (What & Why)
Propagate the Data UI Cheatsheet with the Master Data UI for each station.
2. Verification Method (How)
- [X] Analysis - Code tested by Anthony on 04/8/02020
3. Reviewer (Who & When)
- [X] Within Workgroup Member
4. Location in Tracker
_Determine rules for advancement through kanban defined by checklist._

**Comment by anthonycpichardo** [comment]
Requirements should be chunked out by the tracker header it is in.

Appropriate verification methods when something has been signed off on and moved ahead in the tracker

Everything needs to be verified before it moves ahead in the Kanban

**Comment by lzim** [comment]
@anthonycpichardo and @ritahitching 

_We need to get this Feature "New Issue" template finalized and tested ASAP. Since, it has been the #1 priority on @anthonycpichardo's plate since 3/25 (cross-ref original assignment on #1218), and his responsibility since the end of March and it needs to be closed._
- **Since @anthonycpichardo had not done so, I have edited the Workgroup lab and Assignee to reflect this assignment to him at the 3/25 meeting.**

During an informal huddle with @ritahitching and @staceypark I realized that next steps and a testing routine for verification of the template had not been added.

@ritahitching will create a "new feature" and review the testing checklist to verify whether the following have been down. Consistent with our larger SOP we're recording this on the issue #921.

- @anthonycpichardo I have followed up on this from **2 days ago** and turned it into a testing checklist for @ritahitching that can be tracked by the team:

- [ ] 1. The **use of html to create drop-downs** was approved; but requires instructions for a typical Team PSD member to know exactly where to add their information.

- [ ] 2. Title of each **chunk of the features template** should be named the kanban/tracker name [Delete number 4 and make it the header]

- [ ] 3. Anthony will list the **specific verification methods** that are appropriate for that column of the kanban, which must be signed off on before moving ahead in the tracker.

- [ ] 4. Verification methods for that chunk of the template (aka the column of the tracker) should have **check boxes** that must be checked before advancing in the tracker (like the **manuscript_tracker**).

- [ ] 5. The **issue task cards #s** should be included in the verification list.

**Comment by anthonycpichardo** [comment]
@lzim  Will I be taking ownership of the finalizing the draft that James has produced as well? If so, @jamesmrollins can you add the latest to this thread?

Thanks

**Comment by jamesmrollins** [comment]
Hi @anthonycpichardo the latest is still at the thread location shown below.  I have not updated it (waiting for dust to settle).  Let me know if you need anything.  The Visio diagram in loaded into Lucid charts.

> Hi @anazariz @anthonycpichardo @staceypark @branscombj - Below is the DRAFT Requirements Design and Development Procedure. Feedback encouraged - Enjoy! ~ James
> 

> [req_design_dev_procedure.docx](https://github.com/lzim/teampsd/files/4183549/req_design_dev_procedure.docx)

**Comment by anthonycpichardo** [comment]
@jamesmrollins Thanks! I'll be sure to reach out if I have any questions =).

**Comment by ritahitching** [comment]
@anthonycpichardo some suggestions for the **Feature Template** for the average Team PSD Member.
**Comments:**
I'm not very familiar with HMTL code and it looked a little confusing but I tried to keep them it in. Is it essential?

### **General Guidance for Completing a New Feature Issue Ticket**
1. **Markdown** language to be used, include:
- [ ] **BOLD**
- [ ] _Italic_
- [ ] Use **Numbering**  or **Bullets** 
2.  **HTML** 
- [ ] Text (helper) contained within the <> tags remains unchanged - this is a fixed section header.
- [ ] Description is to be inserted within after the <> arrows for each section of the Feature template
- [ ] To add another requirement copy and paste the helper (Including the \<details>\</details> tags) and paste below the relevant section.
3. **Kanban**  
- [ ] For a **Feature** to progress across the Kanban columns the entire list of requirements within the **Master Card** needs to be checked first
- [ ] If a **Feature** requirement has tasks associated with it cross reference the issue # at the appropriate step.
4. **Guidance for 9 Sidebar Clicks - Delete this text after completing the 9 clicks** 
- [ ]  **Pipeline** - Open Features (Will stay here until card is closed)   
- [ ] **Label**- assign workgroup (lead) responsible for tracking the task  
- [ ] **Projects** - Feature Tracker (work_breakdown)  
- [ ]  **Milestones** - select monthly sprint milestone within the workgroup   
- [ ]  **Estimate** - Estimate feature in complexity points   
- [ ] **Epics** - assign to team-wide epic and TeamPSD master plan project  
- [ ] **Releases** - assign to key team release grouped with other relevant issues  

### **Feature Card Process - Follow All Steps**
1. **Feature Description**
- [ ] Insert brief description answering the following question:  **What is the goal of this feature and why are we working on it?** 
2. **Requirement Definition within < >(What & Why)** 
- [ ] Insert details within after the < > header and use tick box list of requirements.  
- [ ] <Requirement Definition>Requirement Definition</strong> - Review of definition of the requirements.  
- [ ] <strong>Requirement 1</strong> Requirement 1. 
- [ ] <strong>Requirement 2</strong> Requirement 2. 
- [ ] <strong>Requirement 3</strong> Requirement 3. 
3. **Verification Method details within < > (How)**
- [ ] <strong>Documentation of requirements</strong> - Review of definition of the requirements.  
- [ ] <strong>Analysis</strong> - Code, data, model reviewed. 
- [ ] <strong>Test</strong> - Procedures must be well documented.  
- [ ] <strong>Inspect</strong> - Purpose, methods and procedures must be well documented so result can be duplicated using the same inspection procedures. May include inspection of design or an algorithm (aka process map of process).
- [ ] <strong>Demonstrate</strong> - A comprehensive demonstration and/or application of the functions and features of the system or its components.  
- [ ] <strong>Infer</strong> - from “Why”/purpose of requirement definition.  
4. **Reviewer (Who)**
- [ ] <strong>Within Workgroup Member</strong> -Member details
- [ ] <strong>Verify across workgroups</strong> - Workgroups verification
- [ ] <strong>Required or Optional?</strong> -  Option or essential requirement
5. **Requirements, Design, and Test Cases**
- [ ] Outline of the necessary requirements, design, and test steps
- [ ] <strong>Necessary Requirements</strong> - Review of necessary of the requirements.  
6. **Review**: 
- [ ] <strong>Verify List</strong> - Verify that all of the above requirements have been met and signed off on.  
- [ ] <strong>Checkboxes</strong>Once all checkboxes are marked this feature should be moved to the done column.

### **Complete Feature Template Example for Guidance**
1. **Requirement Definition (What & Why) - Must be complete before moving to Operations/Research_To_do**     
  - [X] Community Button links out of Sim UI to the GitHub Community of Practice. 
  - [X] Verification Method (How)   
  - [X] Documentation of requirements - Reviewed definition of the requirements **sim_UI_workgroup** on 4/07/2020.  
2. **Reviewer** 
- [X] Reviewer (Who & When)  
- [X] Verified design with Lindsey & Stacey at **sim_UI_workgroup** on 04/07/2020.
3. **Design Definition (What & Why)  - Must be complete before card moves to **under_development**
- [X] **Create Community Button that links out of Sim UI to the GitHub Community of Practice through a new "I have a question" ["Community   Button" > "I have a question" Button > GitHub "Community of Practice"].**
4. **Verification Method (How)**    
- [X] Design document of the 3 buttons within the Button - Review this design at **sim_UI_workgroup** on 4/14/2020.
5. **Reviewer (Who & When)**  
- [X] Verified design prototype with Lindsey & Stacey at **sim_UI_workgroup** on 04/14/2020.
6. **Build/Test (What & Why)**  
- [X] **Implement the new buttons on the SIM UI** 
7. **Verification Method (How)  - Once build is complete card can move to functional_testing or measurement**
- [X] Analysis - Code tested by Anthony on 04/18/2020.  
- [X] Demonstration of the new design/functionality on the platform - Review this demo at sim_UI_workgroup on 4/21/2020.  
 8. **Reviewer (Who & When)**    
- [X] Anthony ran test scripts outlined in Design Document on 04/18/2020.
- [X] Verified demo with Lindsey & Stacey at **sim_UI_workgroup** on 04/21/2020.

**Comment by lzim** [comment]
**#744 - "Alt 1" vs. "Exp 1" - Considered, but decided to release to DEV.**
- Does it make sense for the Sim UI to continue to use the language of "Alternatives" (e.g., "Alt 1," "Alt 2"), and all our MTL guides use "Experiments" (e.g., "exp 1" "exp 2")?
- Current Run is always black; When you want to pull up the base case, it may be "Alt 2."
- Future uses: In the future, won't be bound to the MTL guides, so exp 1, exp 2 won't be as meaningful as the "Alteratives" defined by specific decisions are included in the name based on our save dialogue.

**DECISIONs re: #744
- Keep use of "alternatives" (not experiments in Control Panel) as it is.**
- Proposal to use "alt a," "alt b" and "alt c." Rejected because it is helpful to use "Alt 1" and Exp 1"
- Base Case - Current Run create problems in changing the Sim UI and are better handled by learner/facilitator.
- Also released for Development

**Comment by anthonycpichardo** [comment]
@lzim Is this comment meant for issue #744? Not sure I'm tracking the addition if it's related to #921

**Comment by anthonycpichardo** [comment]
@branscombj @jamesmrollins @lijenn @staceypark As discussed in yesterday's meeting, can you attempt to create a new feature, using the new feature template [found here](https://github.com/lzim/teampsd/issues/new/choose) and provide any feedback on the process as well as identify any process gaps you may see?

- [ ] Ease of use
- [ ] Steps outlined meet the needs of a feature
- [ ] Clarity as to what the steps represent
- [ ] Any additional feedback you may have

One note: @anazariz  has created a really nice tool to help with the markdown generation for section 2, we will be providing some instruction on it's use in coming meetings and hopefully it will alleviate the pain point I've noted of trying to manage all of the markdown pieces.

**Comment by anthonycpichardo** [comment]
@jamesmrollins Did we come to a decision on how the versioning system will be detailed. From our 4/2 meeting that is the last update remaining. 



[req_design_dev_procedure revised.docx](https://github.com/lzim/teampsd/files/4477216/req_design_dev_procedure.revised.docx)

**Comment by staceypark** [comment]
@anthonycpichardo I would add an instruction to delete sections 1-3 as well not just the ZenHub 9 clicks, before the person posts the card. Alternatively, since these instructions can be helpful as the feature card keeps progressing in the tracker, maybe allowing it to collapse with a dropdown button.

>Notes:
>1. **Markdown** language to be used, include:
> - [ ] **BOLD**
> - [ ] _Italic_
> - [ ] Use **Numbering**  or **Bullets** 
> 

Description is to be inserted after the <> arrows... (take out the word "within")
> 2.  **HTML** 
> - [ ] Text contained within the <> tags remains unchanged - this is a fixed section header.
> - [ ] Description is to be inserted within after the <> arrows for each section of the Feature template
> - [ ] To add another requirement, copy and paste the helper (Including the \<details>\</details> tags) and paste below the relevant section.

> 3. **Kanban**  
> - [ ] For a **Feature** to progress across the Kanban columns the entire list of requirements within the **Master Card** needs to be checked first
> - [ ] If a **Feature** requirement has tasks associated with it cross reference the issue # at the appropriate step.
> 

> **Guidance for 6 Sidebar Clicks - Delete this text after completing the 6 clicks** 
> 
> i. **Pipeline** - Open Features (Will stay here until card is closed)   
> ii. **Label**- assign workgroup (lead) responsible for tracking the task  
> iii. **Projects** - Feature Tracker (work_breakdown)  
> iv. **Milestones** - select monthly sprint milestone within the workgroup   
> v. **Epics** - assign to team-wide epic and TeamPSD master plan project  
> vi. **Releases** - assign to key team release grouped with other relevant issues  


> **1. Feature Description** 
> ------------
> - [ ] Insert brief description answering the following question:  **What is the goal of this feature and why are we working on it?** 
> 
> <details>
> <summary> Requirement Helper </summary>
>  
>  1. Requirement Definition (What & Why)
>  </br>
>  

>  - [ ] --
>  
>  </br>
> 

>2. Verification Method (How)

Is this getting checked off once the requirements are reviewed and approved? If so, I would say "Requirement definitions have been documented and reviewed"

> 
> #### Applicable for Requirements Phase:
> 
> - [ ] **Documentation of requirements** - Review of definition of the requirements.  


Change "Analysis" to "Analyze". All the other checkboxes are verbs. 
Not sure what to do in the "Infer" step. All the other checkboxes are structured as a "title" + sentence explaining title. The "Infer" step seems to be a sentence running on. Also, does not tell me "what" I'm inferring from the purpose of requirement definition.
Algorithm has an "i". 
>  #### Applicable for Design Phase
> - [ ] **Analysis** - Code, data, model reviewed. 
> - [ ] **Infer** - from “Why”/purpose of requirement definition.   
> - [ ] **Inspect** - Purpose, methods and procedures must be well documented so result can be duplicated using the same inspection procedures. May include inspection of design or an algorthm (aka process map of process).  

Again change "Analysis" to "Analyze" for active verb.
> #### Applicable for Build/Test Phase
> - [ ] **Demonstrate** - A comprehensive demonstration and/or application of the functions and features of the system or its components. 
> - [ ] **Test** - Procedures must be well documented. 
> - [ ] **Analysis** - Code, data, model reviewed. 

</br></br>

Why doesn't this section have checkboxes? Are you supposed to write anything down here or is this guidance about reviewing? 

> 3. Reviewer (Who)
> 
> Within Workgroup Member  
> 
> Verify across workgroups  
> 
> Required or Optional?  
> </details>
> 

Not sure how the top half interacts with this bottom half or what the distinction may be. Seems like they have matching stages. Would you be checking of each section in parallel??

> 
> **2. Requirements, Design, and Test Cases: Outline of the necessary requirements, design, and test steps**  
> -------------
> 
> 
> - [ ] Requirement 1 Met
> <details >
> <summary>Requirement 1 -Rename to "What" of this requirement</summary>
> <div>
>  
>   1. **Requirement Definition (What & Why)** - Must be complete before moving to Operations/Research_To_do     
>   Example:  
>   - [X] Community Button links out of Sim UI to the GitHub Community of Practice. 
>   2. **Verification Method (How)**   
>   Example:  
>   - [X] Documentation of requirements - Reviewed definition of the requirements at **sim_UI_workgroup** on 4/07/2020.  
>   3. **Reviewer (Who & When)**  
>   Example:  
>   - [x] Verified design with Lindsey & Stacey at **sim_UI_workgroup** on 04/07/2020.
> 
> 
>   ---
>   1. **Design Definition (What & Why)**  - Must be complete before card moves to under_development
>    Example:  
>   - [X] Create Community Button that links out of Sim UI to the GitHub Community of Practice through a new "I have a question" ["Community   Button" > "I have a question" Button > GitHub "Community of Practice"].
>   2. **Verification Method** (How)  
>    Example:  
>   - [ ] Design document of the 3 buttons within the Button - Review this design at **sim_UI_workgroup** on 4/14/2020.
>   3. **Reviewer** (Who & When)  
>    Example:    
>   - [ ] Verified design prototype with Lindsey & Stacey at **sim_UI_workgroup** on 04/14/2020.
>   
> 
>   ---
>   1. **Build/Test** (What & Why)  
>   Example:  
>   - [X] Implement the new buttons on the SIM UI
>   2. **Verification Method** (How)  - Once build is complete card can move to functional_testing or measurement
>   Example:  
>   - [X] Analysis - Code tested by Anthony on 04/18/2020.  
>   - [ ] Demonstration of the new design/functionality on the platform - Review this demo at **sim_UI_workgroup** on 4/21/2020.  
>   3. **Reviewer** (Who & When)    
>    Example:  
>   - [X] Anthony ran test scripts outlined in Design Document on 04/18/2020.
>   - [ ] Verified demo with Lindsey & Stacey at **sim_UI_workgroup** on 04/21/2020.
>  </div>
> </details>
> 
> *Note: To add another requirement copy and paste the helper above (Including the \<details>\</details> tags) and paste below.*
> 
> 
> **3. Review**: 
> ----------------
> ### Verify that all of the above requirements have been met and signed off on.
>   
> #### Once all checkboxes are marked this feature should be moved to the done column.
>

**Comment by anthonycpichardo** [comment]
@staceypark Thanks for the feedback!

> Not sure how the top half interacts with this bottom half or what the distinction may be. Seems like they have matching stages. Would you be checking of each section in parallel??

@staceypark  Are you referring to the " Requirements Helper " and The requirements section? I can add a clarification that the helper is meant to be used a tool to help provide structure for the requirements.

**Comment by jamesmrollins** [comment]
@anthonycpichardo  We did not come to any conclusions about versioning.  I was going to take a crack at it, but never got to it.  The key question is how versioning affects the product line.  This may affect the study side of the house, since experimental groups under different version may affect the veracity of the science.  Thus, this is a @lzim PI call.

**Comment by jamesmrollins** [comment]
@anthonycpichardo  I used this on a real case to start a new feature. In general, the instructions provided are too cryptic for me to understand what is required to fill in this form properly.  Let me know if you have any questions. I agree with @staceypark regarding all the notes in the header.  In some respects, I wonder if we should expect people know markdown to use this. My feedback follows:

Regarding the Description and Goal:
![image](https://user-images.githubusercontent.com/30132017/79284346-39897200-7e6f-11ea-8a78-f420d17738ec.png)

Regarding Requirements:
![image](https://user-images.githubusercontent.com/30132017/79284615-0b586200-7e70-11ea-90b1-ddc4872498c9.png)

Regarding Verification Method:
![image](https://user-images.githubusercontent.com/30132017/79285102-60e13e80-7e71-11ea-826c-cfa576e1731d.png)

**Comment by branscombj** [comment]
Well, I get an F on this assignment. So sorry; I just couldn't get a foothold. My GitHub skill is very tenuous. I could use some background on who is envisioned to use it, because it may be fine if it's for TeamPSD, particularly HQ. Others of us too, with training.

**Comment by anthonycpichardo** [comment]
> Well, I get an F on this assignment. So sorry; I just couldn't get a foothold. My GitHub skill is very tenuous. I could use some background on who is envisioned to use it, because it may be fine if it's for TeamPSD, particularly HQ. Others of us too, with training.

Thanks for the feedback =)! We definitely want this to be usable but it seems some training may be needed or some restructuring of how it's laid out. Ash built a really nice tool to help with the markdown generation that may provide an assist on a large bit of this. We will introduce it shortly.

Thanks @branscombj !!

**Comment by lijenn** [comment]
@branscombj I edited your comment to cross reference your test feature card as you've provided lots of feedback there and I'm not sure if anyone saw it. 

@anthonycpichardo Seems like James's first 2 feedback comments regarding Description and Goal were similar to the confusion I had regarding how the card was progressing/where it was staying and the confusion between the card's placement in the Feature Tracker or ZenHub Board since the Pipelines can be easily confused with the Kanban Columns in the Feature Tracker (e.g. both the Feature Tracker and ZenHub Board have a section for cards considered "Done". And whether or not this was serving as the "Master" feature card that would lead to breakdown of "task" cards. May want to be very explicit and mundane about explaining this. 
**I think this may be the most important part to be clear about since it regards following SOP.**

I also agree @jamesmrollins assumption of people needing a good understanding or at least the intuition of the syntax for markdown. I know we had markdown training prior, but when I first initially reviewed this, it helped me to understand it and see it more easily by referring to the 'Preview' of the card first. 
I look forward to testing Ash's tool for section 2 since I also think that it looks a bit complex right now and see if we could include it future markdown training documents as well. 

With that said:
Ease of use: markdown format makes it look difficult to use, especially section 2
Clarity as to what the steps represent: the purpose/procedure of listing the steps itself is pretty clear to me if that's what you meant by this
Steps outlined meet the needs of a feature: Looks good to me!
Comments: Flow of how the card works could better reflect the flow of the SOP

**Comment by lzim** [comment]
> > Well, I get an F on this assignment. So sorry; I just couldn't get a foothold. My GitHub skill is very tenuous. I could use some background on who is envisioned to use it, because it may be fine if it's for TeamPSD, particularly HQ. Others of us too, with training.
> 
> Thanks for the feedback =)! We definitely want this to be usable but it seems some training may be needed or some restructuring of how it's laid out. Ash built a really nice tool to help with the markdown generation that may provide an assist on a large bit of this. We will introduce it shortly.
> 
> Thanks @branscombj !!

@anazariz We plan to review your Markdown tool at today's **3PM Support Workgroup Leads meeting,** But, since #921 is running so late, and there is a lot of rich feedback [THANKS TEAM 🏆!!}  here about making this template a) **clearer** and b) **SOP consistent** ... 
- _Is there any way you can debut your Markdown tool asynchronously online?_ 
- It would provide a more robust test of how we would need to use it anyway, and it would keep this moving online today so we can get to done ASAP (keep in mind some team members are 3 hours ahead of us).

FYI: @anthonycpichardo @staceypark @jamesmrollins @branscombj @lijenn

**Comment by lzim** [comment]
**#921 Next Steps for **master_feature_card** template:**

**Key orienting question: To what extent the complexity .xlsm confers advantages over #1044 existing checklist knowledge and plain English?**

1. @lzim is interested in reviewing for language consistency with the feature_tracker

Preferred steps for each requirement: DEFINE, DESIGN (freeze) DEVELOP, TEST, PRODUCE

2. @anthonycpichardo will draft list of valid verification methods for each of the 5 steps above.

3. **quant_workgroup** - Articulate the tradeoffs from your perspective of complexity vs. simplicity.

4. Try online asynchronous test with other Team PSD members.

**Comment by lzim** [comment]
**DECIDED during 4/15/2020 Support Workgroup Meeting:**
#921 - "Lindsey, how do we want to handle versioning between the SIM and MTL?"

- Releases (aka Versions) have to be tracked carefully so we know what version of each wave of our trials learners received and because we need to fully document and train others before we release something to the field.

- Lindsey: Not sure what "Sim" and "_MTL_" mean in this use.

**DECISIONS: We already decided we would not release new "releases" (version) other than when a new wave starts (~8-9 months).

**Comment by lzim** [comment]
**Decisions made at **quant_workgroup** meeting on 4/16/2020:**

1. Edited the "issue" template (per #hqhuddle on 4/15/2020)
2. Edited the "feature" template for Team testing
3. Removed the "measurement" column of the **feature_tracker**
- [ ] @anthonycpichardo will remove reference to this in the SOP.
- [ ] @anthonycpichardo will make the remaining edits from 4/15/2020 **Support Workgroup Leads** meeting.

** Once the SOP edits are made, we'll close #921 😅

**Comment by anthonycpichardo** [comment]
I've made the edits to the SOP and added the Design Review Development Policy and Process to the wiki. Ready to close this issue.

@lzim  @jamesmrollins  @staceypark  @anazariz
